python/ray/tune/progress_reporter.py

```diff
@@ -14,67 +14,166 @@
                       "Please re-run 'pip install ray[tune]' or "
                       "'pip install ray[rllib]'.")
 
-DEFAULT_PROGRESS_KEYS = DEFAULT_RESULT_KEYS + (EPISODE_REWARD_MEAN, )
-# Truncated representations of column names (to accommodate small screens).
-REPORTED_REPRESENTATIONS = {
-    EPISODE_REWARD_MEAN: "reward",
-    MEAN_ACCURACY: "acc",
-    MEAN_LOSS: "loss",
-    TIME_TOTAL_S: "total time (s)",
-    TIMESTEPS_TOTAL: "timesteps",
-    TRAINING_ITERATION: "iter",
-}
-
 
 class ProgressReporter:
-    # TODO(ujvl): Expose ProgressReporter in tune.run for custom reporting.
+    def should_report(self, trials):
+        """Returns whether or not progress should be reported.
+
+        This is called at the end of every execution step.
+        """
+        raise NotImplementedError
 
-    def report(self, trial_runner):
-        """Reports progress across all trials of the trial runner.
+    def report(self, trials, *sys_info):
+        """Reports progress across trials.
 
         Args:
-            trial_runner: Trial runner to report on.
+            trials (list[Trial]): Trials to report on.
+            sys_info: System info.
         """
         raise NotImplementedError
 
 
-class JupyterNotebookReporter(ProgressReporter):
-    def __init__(self, overwrite):
+class TuneReporterBase(ProgressReporter):
+    """Base class for the default Tune reporters."""
+
+    # Truncated representations of column names (to accommodate small screens).
+    DEFAULT_COLUMNS = {
+        EPISODE_REWARD_MEAN: "reward",
+        MEAN_ACCURACY: "acc",
+        MEAN_LOSS: "loss",
+        TIME_TOTAL_S: "total time (s)",
+        TIMESTEPS_TOTAL: "timesteps",
+        TRAINING_ITERATION: "iter",
+    }
+
+    def __init__(self,
+                 metric_columns=None,
+                 max_progress_rows=20,
+                 max_error_rows=20):
+        """Initializes a new TableReporterBase.
+
+        Args:
+            metric_columns (dict[str, str]|list[str]): Names of metrics to
+                include in progress table. If this is a dict, the keys should
+                be metric names and the values should be the displayed names.
+                If this is a list, the metric name is used directly.
+            max_progress_rows (int): Maximum number of progress rows to print
+                in the progress table. Defaults to 20.
```

Comment:
As a reader, can you help clarify what is a progress row as compared to an error row?